### PR TITLE
Added Halokeys Elemental 75 Keyboard

### PR DESCRIPTION
**Pull request by shamit05**
This commit adds the Elemental 75 keyboard produced by Halokeys.

## Description

Added QMK support to Elemental 75. Provides two different keymaps: default and via. Keyboard uses custom PCB built on STM32F303 microcontroller. 

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/halokeys/elemental75/config.h [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+#define DIODE_DIRECTION COL2ROW
+#define DEBOUNCE 5
+
+#define ENCODERS_PAD_A { B6 }
+#define ENCODERS_PAD_B { B5 }
+#define ENCODER_RESOLUTION 2
+#define TAP_CODE_DELAY 10 // tap_code function delay for register and unregister
+
+#define RGB_DI_PIN A10
+#define RGBLIGHT_ANIMATIONS
```
Comment (drashna) [inline keyboards/halokeys/elemental75/config.h:42]: Could you explicitly define which animations are enabled? 
```suggestion
#define RGBLIGHT_EFFECT_BREATHING
#define RGBLIGHT_EFFECT_RAINBOW_MOOD
#define RGBLIGHT_EFFECT_RAINBOW_SWIRL
#define RGBLIGHT_EFFECT_SNAKE
#define RGBLIGHT_EFFECT_KNIGHT
#define RGBLIGHT_EFFECT_CHRISTMAS
#define RGBLIGHT_EFFECT_STATIC_GRADIENT
#define RGBLIGHT_EFFECT_RGB_TEST
#define RGBLIGHT_EFFECT_ALTERNATING
#define RGBLIGHT_EFFECT_TWINKLE
```

File: keyboards/halokeys/elemental75/elemental75.c [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+#include "elemental75.h"
+#include "print.h"
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+  /* If console is enabled, it will print the matrix position and status of each key pressed */
+  #ifdef CONSOLE_ENABLE
+      uprintf("KL: kc: 0x%04X, col: %u, row: %u, pressed: %b, time: %u, interrupt: %b, count: %u\n", keycode, record->event.key.col, record->event.key.row, record->event.pressed, record->event.time, record->tap.interrupted, record->tap.count);
+  #endif
+    return true;
+}
```
Comment (drashna) [inline keyboards/halokeys/elemental75/elemental75.c:25]: This should be the kb level function, not the user level.  

And if it's the kb level, then it needs to to call the user level, as well.

File: keyboards/halokeys/elemental75/elemental75.json [hunk 3]
```diff
@@ -0,0 +1,375 @@
+{
```
Comment (drashna) [inline keyboards/halokeys/elemental75/elemental75.json:1]: This should be removed

File: keyboards/halokeys/elemental75/rules.mk [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+MCU = STM32F303
+BOARD = QMK_PROTON_C
+
+# Bootloader selection
+BOOTLOADER = stm32-dfu
+
+# Build Options
+#   change yes to no to disable
+#
+BOOTMAGIC_ENABLE = lite     # Virtual DIP switch configuration
```
Comment (drashna) [inline keyboards/halokeys/elemental75/rules.mk:11]: There is no "full" bootmagic anymore, only the "lite" option, which can be turned on with just "yes".
```suggestion
BOOTMAGIC_ENABLE = yes      # Enable Bootmagic Lite
```

File: keyboards/halokeys/elemental75/rules.mk [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+MOUSEKEY_ENABLE = yes        # Mouse keys
+EXTRAKEY_ENABLE = yes       # Audio control and System control
+CONSOLE_ENABLE = no         # Console for debug
+COMMAND_ENABLE = no         # Commands for debug and configuration
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = yes           # USB Nkey Rollover
```
Comment (drashna) [inline keyboards/halokeys/elemental75/rules.mk:19]: Please remove the `SLEEP_LED_ENABLE` setting, as it's not really used, and remove the comments for both it and NKRO.  
```suggestion
NKRO_ENABLE = no            # Enable N-Key Rollover
```

File: keyboards/halokeys/elemental75/rules.mk [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE = yes       # Enable keyboard RGB underglow
+BLUETOOTH_ENABLE = no       # Enable Bluetooth
```
Comment (drashna) [inline keyboards/halokeys/elemental75/rules.mk:22]:
```suggestion
```

File: keyboards/halokeys/elemental75/rules.mk [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+AUDIO_ENABLE = no           # Audio output
+ENCODER_ENABLE = yes
+#LAYOUTS = ortho_5x15
+TAP_DANCE_ENABLE = yes
```
Comment (drashna) [inline keyboards/halokeys/elemental75/rules.mk:26]: Tap Dance should not be enabled at the keyboard level, as it requires user configuration to work properly.   It should only be enabled at the keymap level. 

```suggestion
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/halokeys/elemental75/keymaps/default/keymap.c [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+      if (clockwise) {
+          print("Tapped code: KC_VOLU\n");
+          tap_code(KC_VOLU);
+      } else {
+          print("Tapped code: KC_VOLD\n");
+          tap_code(KC_VOLD);
+      }
+    }
+    return false;
+}
```
Comment (drashna) [inline keyboards/halokeys/elemental75/keymaps/default/keymap.c:88]:
```suggestion
```

File: keyboards/halokeys/elemental75/keymaps/via/keymap.c [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+      if (clockwise) {
+          print("Tapped code: KC_VOLU\n");
+          tap_code(KC_VOLU);
+      } else {
+          print("Tapped code: KC_VOLD\n");
+          tap_code(KC_VOLD);
+      }
+    }
+    return false;
+}
```
Comment (drashna) [inline keyboards/halokeys/elemental75/keymaps/via/keymap.c:103]:
```suggestion
```

File: keyboards/halokeys/elemental75/elemental75.c [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+  * This program is distributed in the hope that it will be useful,
+  * but WITHOUT ANY WARRANTY; without even the implied warranty of
+  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  * GNU General Public License for more details.
+  *
+  * You should have received a copy of the GNU General Public License
+  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+  */
+#include "elemental75.h"
+#include "print.h"
```
Comment (drashna) [inline keyboards/halokeys/elemental75/elemental75.c:17]:
```suggestion
#ifdef ENCODER_ENABLE
bool encoder_update_kb(uint8_t index, bool clockwise) {
    if (!encoder_update_user(index, clockwise)) {
        return false;
    }
    uprintf("Encoder Hit: clockwise: %b, index: %u\n", clockwise, index);
    if (index == 0) {
      if (clockwise) {
          print("Tapped code: KC_VOLU\n");
          tap_code(KC_VOLU);
      } else {
          print("Tapped code: KC_VOLD\n");
          tap_code(KC_VOLD);
      }
    }
    return true;
}
#endif
```

**Review by drashna — CHANGES_REQUESTED** [review]

**Review by drashna — APPROVED** [review]

File: keyboards/halokeys/elemental75/info.json [hunk 11]
```diff
@@ -0,0 +1,100 @@
+{
+    "keyboard_name": "Elemental 75",
+    "url": "",
+    "maintainer": "shamit05",
+    "layout_aliases": {
+        "LAYOUT": "LAYOUT_all"
+    },
```
Comment (fauxpark) [inline keyboards/halokeys/elemental75/info.json:7]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/halokeys/elemental75/chconf.h [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
```
Comment (zvecr) [inline keyboards/halokeys/elemental75/chconf.h:16]:
```suggestion
#pragma once

```

**Review by zvecr — CHANGES_REQUESTED** [review]

**Review by drashna — APPROVED** [review]

**Comment by shamit05** [comment]
Can someone take a look at this? Thanks.

**Review by fauxpark — APPROVED** [review]

File: keyboards/halokeys/elemental75/keymaps/default/keymap.c [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+
+/* Tap Dance declarations */
+enum {
+    TD_PLAY_FORWARD_BACK,
+};
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+  /* If console is enabled, it will print the matrix position and status of each key pressed */
+  #ifdef CONSOLE_ENABLE
+      uprintf("KL: kc: 0x%04X, col: %u, row: %u, pressed: %b, time: %u, interrupt: %b, count: %u\n", keycode, record->event.key.col, record->event.key.row, record->event.pressed, record->event.time, record->tap.interrupted, record->tap.count);
```
Comment (drashna) [inline keyboards/halokeys/elemental75/keymaps/default/keymap.c:33]:
```suggestion
      uprintf("KL: kc: 0x%04X, col: %u, row: %u, pressed: %d, time: %u, interrupt: %d, count: %u\n", keycode, record->event.key.col, record->event.key.row, record->event.pressed, record->event.time, record->tap.interrupted, record->tap.count);
```

File: keyboards/halokeys/elemental75/keymaps/via/keymap.c [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+
+/* Tap Dance declarations */
+enum {
+    TD_PLAY_FORWARD_BACK,
+};
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+  /* If console is enabled, it will print the matrix position and status of each key pressed */
+  #ifdef CONSOLE_ENABLE
+      uprintf("KL: kc: 0x%04X, col: %u, row: %u, pressed: %b, time: %u, interrupt: %b, count: %u\n", keycode, record->event.key.col, record->event.key.row, record->event.pressed, record->event.time, record->tap.interrupted, record->tap.count);
```
Comment (drashna) [inline keyboards/halokeys/elemental75/keymaps/via/keymap.c:30]:
```suggestion
      uprintf("KL: kc: 0x%04X, col: %u, row: %u, pressed: %d, time: %u, interrupt: %d, count: %u\n", keycode, record->event.key.col, record->event.key.row, record->event.pressed, record->event.time, record->tap.interrupted, record->tap.count);
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/halokeys/elemental75/config.h [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+#define MATRIX_COLS 15
+
+#define MATRIX_ROW_PINS { A2, A3, A4, A5, A6, A7 }
+#define MATRIX_COL_PINS { B2, B1, B0, B10, B11, B13, B14, B15, A8, A9, A14, A15, B3, B4, B7 }
+#define DIODE_DIRECTION COL2ROW
+#define DEBOUNCE 5
+
+#define ENCODERS_PAD_A { B6 }
+#define ENCODERS_PAD_B { B5 }
+#define ENCODER_RESOLUTION 2
```
Comment (drashna) [inline keyboards/halokeys/elemental75/config.h:38]: Not a fan of this being set by default. 

```suggestion
```

File: keyboards/halokeys/elemental75/elemental75.c [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+  */
+#include "elemental75.h"
+
+#ifdef ENCODER_ENABLE
+bool encoder_update_kb(uint8_t index, bool clockwise) {
+    if (!encoder_update_user(index, clockwise)) {
+        return false;
+    }
+    uprintf("Encoder Hit: clockwise: %b, index: %u\n", clockwise, index);
```
Comment (drashna) [inline keyboards/halokeys/elemental75/elemental75.c:23]:
```suggestion
    uprintf("Encoder Hit: clockwise: %d, index: %u\n", clockwise, index);
```

**Review by fauxpark — APPROVED** [review]

**Comment by shamit05** [comment]
Can someone revise the most recent changes and merge if it's good? Thanks!

**Review by drashna — APPROVED** [review]

**Pull request merged**